### ISSUE TITLE
[BUG] Flash Attention TileLang kernel compilation fails on RTX 5060 Ti (Blackwell    architecture)

### ISSUE THREAD

**Issue by Virtuoso461** [tile-ai/tilelang#1871]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

ileLang: 0.1.8 (likely too old for RTX 5060 Ti support)      

### System information

  Versions:                                                                          
  - TileLang: 0.1.8 (likely too old for RTX 5060 Ti support)                         
  - PyTorch: 2.10.0                                                                  
  - CUDA: 13.1                                                                       
  - GPU: RTX 5060 Ti  

### Problem description

Title: Flash Attention TileLang kernel compilation fails on RTX 5060 Ti (Blackwell 
  architecture)                                                                      
                                                                                     
  Description:                                                                       
  TileLang's Flash Attention kernel fails to compile during inference on RTX 5060 Ti 
  GPUs. The kernel build fails with Failed to build prim_func from main error, but   
  the system falls back to PyTorch implementation, allowing inference to continue    
  with reduced performance.                                                          
                                                                                     
  Environment:                                                                       
  - GPU: NVIDIA RTX 5060 Ti (Blackwell architecture)                                 
  - CUDA Version: [请运行 nvidia-smi 查看]                                           
  - Driver Version: [请运行 nvidia-smi 查看]                                         
  - Compute Capability: [请运行 nvidia-smi --query-gpu=compute_cap --format=csv 查看]
  - OS: Linux (Ubuntu)                                                               
  - Python: 3.12                                                                     
  - PyTorch: [请运行 pip show torch | grep Version 查看]                             
  - TileLang: [请运行 pip show tilelang | grep Version 查看]                         
                                                                                     
  Steps to Reproduce:                                                                
  1. Train a 125M model with TileLang kernels enabled                                
  2. After training, run inference:                                                  
  python examples/generate.py \                                                      
      --checkpoint checkpoints/checkpoint-10000 \                                    
      --prompt "韩立修炼" \                                                          
      --max_new_tokens 100                                                           
  3. Observe the CRITICAL error messages during generation                           
                                                                                     
  Expected Behavior:                                                                 
  Flash Attention kernel should compile successfully and run efficiently on RTX 5060 
  Ti.                                                                                
                                                                                     
  Actual Behavior:                                                                   
  Multiple CRITICAL errors appear:                                                   
  [TileLang:tilelang.language.eager.builder:CRITICAL]: Failed to build prim_func from
   main                                                                              
                                                                                     
  However, inference continues using fallback implementation.                        
                                                                                     
  Error Log:                                                                         
  2026-02-23 23:29:17  [TileLang:tilelang.language.eager.builder:CRITICAL]: Failed to
   build prim_func from main                                                         
  args={'Q': Q, 'K_cache': K_cache, 'V_cache': V_cache, 'Output': Output}            
  source=def make_closure(accum_dtype, batch, block_n, cache_seq_len, dtype,         
  head_dim, n_heads, num_kv_heads, scale):                                           
      ...                                                                            
                                                                                     
  Key Observations:                                                                  
  1. Kernel compilation fails repeatedly during inference loop                       
  2. Model has Engram module enabled at layer 1                                      
  3. Using bfloat16 dtype                                                            
  4. KV cache parameters: cache_seq_len, block_n                                     
  5. Error occurs in flash_attn.py at line 371-455                                   
  6. System has fallback mechanism and continues with PyTorch native implementation  
                                                                                     
  Impact:                                                                            
  - Training continues but without optimized Flash Attention                         
  - Inference works but with ~20-30% performance degradation                         
  - Higher memory usage due to lack of optimization                                  
                                                                                     
  Workaround:                                                                        
  Disabling TileLang with --no_use_tilelang flag allows training/inference to proceed
   using PyTorch native attention:                                                   
  python examples/train.py --config 125M --no_use_tilelang ...                       
                                                                                     
  Possible Root Causes:                                                              
  1. RTX 5060 Ti uses Blackwell architecture (compute capability 9.x), which may not 
  be fully supported yet                                                             
  2. TileLang version may be outdated and lack Blackwell support                     
  3. CUDA version compatibility issue (5060 Ti requires CUDA 12.6+)                  
  4. BF16 data type handling in the kernel compilation                               
                                                                                     
  Additional Context:                                                                
  - RTX 5060 Ti is a very new GPU architecture (released 2024)                       
  - Training loss decreases normally (11.7 → 1.8), suggesting model architecture is  
  correct                                                                            
  - Issue appears during both training and inference when KV cache is used           
  - The kernel tries to use pipelined loops and shared memory optimization           
                                                                                     
  ---  

### Reproducible example code

The Python snippets:

```python

```

### Traceback

```pytb

```

**Comment by LeiWang1999** [comment]
Would you mind providing detailed scripts to help us reproduce?

**Comment by Hale423** [comment]
Seems like fmha realization in blackwell is empty for now, the existed [examples ](https://github.com/tile-ai/tilelang/tree/main/examples/flash_attention) are from hopper archs

**Comment by Hale423** [comment]
@LeiWang1999  Can I pick this up for now,  I suppose the pipeline that leverage blackwell features should be like this:
```
Loop over K/V blocks:
  1. GEMM SS:  S = Q_shared @ K_shared^T → S_tmem (fp32)     
  2. tcgen05.ld: S_tmem → S_reg                               
  3. Online softmax on registers → P_reg (cast to bf16)       
  4. Rescale: O_tmem → O_reg, O_reg *= scale, O_reg → O_tmem [tcgen05.ld/st fp32]
  5. tcgen05.st: P_reg → P_tmem (bf16 packed)                 [PR #1866]
  6. GEMM TS:  O_tmem += P_tmem @ V_shared                    [PR #1866]
End loop
tcgen05.ld: O_tmem → O_reg, normalize, copy out
```

**Comment by LeiWang1999** [comment]
@Hale423 Sure, please have a try, and we're also working on a pass to automatically inject tmem copy.

**Comment by Hale423** [comment]
Got it! 
I'll start with a manual Blackwell FMHA example using the tcgen05.st + gemm_ts primitives, Maybe serve as a reference/test case for the auto-inject pass you're working on.

**Comment by iori2333** [comment]
@Hale423 #1910 is built for Blackwell (SM100). However, architecture of RTX 50xx is [SM120](https://developer.nvidia.com/cuda/gpus), which does not support tcgen05 ptx, although they share same name Blackwell.

**Comment by Hale423** [comment]
> [@Hale423](https://github.com/Hale423) [#1910](https://github.com/tile-ai/tilelang/pull/1910) is built for Blackwell (SM100). However, architecture of RTX 50xx is [SM120](https://developer.nvidia.com/cuda/gpus), which does not support tcgen05 ptx, although they share same name Blackwell.

Oh I realized the differences, will implement the examples and verify on RTX 50x0 devices, thanks for reminding  =。=

**Comment by Virtuoso461** [comment]
> [#1910](https://github.com/tile-ai/tilelang/pull/1910)是为Blackwell（SM100）制造的。然而，RTX 50xx 的架构是 [SM120](https://developer.nvidia.com/cuda/gpus)，不支持 tcgen05 ptx，尽管两者同名 Blackwell。

The SM120 architecture requires calling methods for mma (matrix multiply-accumulate) without the tcgen05 prefix, while the SM100 architecture uses the tcgen05 prefix. Is my understanding correct? If so, I roughly have an idea of how to optimize and replace the computations within.

**Comment by iori2333** [comment]
@Virtuoso461 Yes, [tcgen05.mma](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#tcgen05-mma-instructions) is marked as `sm100a` instruction, which is not compatible with sm120. It seems that [mma.sync](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#warp-level-matrix-instructions-mma) is the only available instruction for MMA on SM120.

**Comment by Hale423** [comment]
@Virtuoso461  Implementing flash attention kernels [examples ](https://github.com/tile-ai/tilelang/tree/main/examples/flash_attention)  on RTX 5080 on my side was okey,  perhaps rebase the code could address the issue.
SM120 uses sync.mma  (the way Ampere does) without Tmem enabled,  I guess the existed examples should be sufficient.

**Comment by Virtuoso461** [comment]
> [@Virtuoso461](https://github.com/Virtuoso461) Implementing flash attention kernels [examples ](https://github.com/tile-ai/tilelang/tree/main/examples/flash_attention) on RTX 5080 on my side was okey, perhaps rebase the code could address the issue. SM120 uses sync.mma (the way Ampere does) without Tmem enabled, I guess the existed examples should be sufficient.

I checked the code repository, and indeed there are no new commits specifically for SM120 yet. You're right on target; the publicly available ones are still mainly the SM100 version from three days ago. The point you mentioned about SM120 using `sync.mma` without Tmem enabled is crucial. This precisely indicates that SM120's instruction path might be closer to Ampere's style, showing a clear difference from SM100 which relies on TMA and WGMMA. Since you're able to run the existing flash attention kernel without issues on your side, it should be safe to continue using the current Ampere-style examples for now.
